### PR TITLE
Neighbor lists optimizations and start of PotentialMaster cleanup

### PR DESCRIPTION
**Pull request by akofke**
- Improve how neighbor lists are calculated from neighbor cells. The current ApiAACell computes cell
neighbors on the fly and is very slow. Now the RectangularLattice computes and caches the
neighboring sites of each site, and the NeighborListManager uses for-loops over the cells and their 
neighbors inline. Lattice neighbors are updated when the lattice size or neighbor range (given as a
double radius). There's still some tangling between RectanguarLattice and CellLattice, and they should probably be merged into one class at some point.

- Other performance improvements to NeighborListManager. Neighbor updates now execute in parallel
using a ForkJoinPool (via parallel streams). The lattice cells are split up among threads and each thread updates the neighbors for all atoms in a cell. Additionally, the NeighborListManager only computes up neighbors by default, until the ensureDownLists() method is called where it constructs the down lists and maintains them from that point on (it is the responsibility of the PotentialMaster to call this method when it receives a downlist calculation request).

- PotentialMasterNbr class hierarchy is changed. PotentialMasters Cell and Site are now their own
separate subclasses of Nbr and are refactored to use new NeighborIterator interface instead of
Api1ACell/Site.

- PotentialMasterList's dependencies are made less complicated. NeighborListManager now creates and holds its own NeighborCellManager.

-  NeighborCriterion are changed to accept two IAtom parameters instead of an AtomList.

### CODE REVIEW

**Comment by ajschult** [comment]
> Neighbor updates now execute in parallel using a ForkJoinPool (via parallel streams). 

What is the default behavior here?  Does it make a thread for every core?

**Comment by akofke** [comment]
It uses the ForkJoinPool static common pool, which by default uses the same number of threads as
processors (main thread + e.g. 3). This can be changed with the [java.util.concurrent.ForkJoinPool.common.parallelism](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html) system property, either using the -D command line switch or a static System.setProperty. If we want more customization at some point (like how much tasks get split) we could use a custom pool and ForkJoinTasks instead of parallel streams (although I benchmarked this and it doesn't make it any faster).

**Comment by ajschult** [comment]
Right, that seems to be broken in the 3 most common use cases.

1. A job on CCR requesting 1 task is going to try to use 10-30 threads, which will all compete for one core.  It's likely that neighbor listing will be noticeably slower.
2. A job on our workstations is going to try to use 10-20 threads, even though it's supposed to use only one core.  Other jobs on the same node will suffer trying to compete.
3. A user sitting at a workstation, running a simulation in intellij is going to use 10-20 threads.  Any jobs running on the workstation are going to lose out (the jobs run with the lowest priority).

Cases where threads will help now:
1. User runs one simulation on their laptop
2. User runs no more than 2 simulations on their workstation while their workstation is not running jobs.

Both of these are pretty rare.

If we can get multithreading to the point that we have good parallel efficiency, then it will make sense to run multicore jobs on CCR, but my understanding is that we are not there.  In the meantime, users in situations where multithreading will help can enable it, but users should not have to jump through hoops to disable it for the most common cases (because... they won't).

**Comment by akofke** [comment]
Ok, the multithreading can be disabled by default for now and you have to opt in with the "etomica.nbr.parallel" system property.

File: etomica-core/src/main/java/etomica/nbr/list/NeighborListManager.java [hunk 1]
```diff
@@ -417,7 +416,7 @@ public void addAtomNotify(IAtom atom) {
 
     public double getRange() {
         // TODO
```
Comment (ajschult) [inline etomica-core/src/main/java/etomica/nbr/list/NeighborListManager.java]: Is this still TODO?

**Comment by ajschult** [comment]
Refactor PML doNBodyStuff -- "Pre-sizing AtomArrayList avoids internal array reallocations..."

"clear" only sets itemsInList to 0.  The array is only re-allocated if the new size is bigger than the existing array length (not itemsInList).  Reallocation happens a few times at the beginning of the simulation and then never again.

**Comment by akofke** [comment]
Ah true. Since we're considering moving the n-body stuff to its own PotentialMaster we can probably just keep it like this for now (since it means one less field).

File: etomica-core/src/main/java/etomica/nbr/list/NeighborListManager.java (hunk 1 above)
Comment (akofke) [reply etomica-core/src/main/java/etomica/nbr/list/NeighborListManager.java]: If we're satisfied that the whole setRange stack (pml -> nlm -> ncm -> lattice) works correctly then no.

Comment (akofke) [reply etomica-core/src/main/java/etomica/nbr/list/NeighborListManager.java]: This is probably a good thing to test in NeighborListingTest - what's a reasonable value I can change it to that would result in different neighbors?

**Comment by ajschult** [comment]
http://rheneas.eng.buffalo.edu/tinderbox/tinderbox/showbuilds.cgi?tree=master
http://rheneas.eng.buffalo.edu/tinderbox/showbuilds.cgi?tree=neighbor-lists-optimizations

```
SWC     500 75 => 75      4000   116 => 109
HS      500 37 => 36      4000    56 => 50     32000  86 => 84
LJMC    500 50 => 50      4000    54 => 64
LJGCMC  500 46 => 50      4000    53 => 68
LJMD    500 61 => 59      4000    77 => 73
```

MD has a modest perf boost (0-11%) and MC has a modest perf hit (0-28%).  NeighborListManager itself is unused by MC.  Is the perf hit then from Api1ACell?  Or just general PM restructuring?  I'm not sure what else from this would impact MC.

**Comment by akofke** [comment]
Hmm that doesn't seem right. MC is definitely affected (PotentialMasterCell is mostly rewritten) but the JMH benchmarks show an improvement. Ignore LJMD3D, the benchmark changed between branches.

master
```
Benchmark                                  (nCellsZ)  (numMolecules)  (numSteps)   Mode  Cnt      Score       Error  Units
BenchSimCatalysis.integratorStep                  20             N/A         N/A  thrpt    3   3533.346 ±   490.563  ops/s
BenchSimDCVGCMD.integratorStep                   N/A             N/A         N/A  thrpt    5   1571.728 ±    83.091  ops/s
BenchSimHSMD3D.integratorStep                    N/A             500    20000000  thrpt    5   4392.060 ±   107.071  ops/s
BenchSimHSMD3D.integratorStep                    N/A            4000    20000000  thrpt    5    428.471 ±    39.539  ops/s
BenchSimHSMD3D.integratorStep                    N/A           32000    20000000  thrpt    5     37.111 ±     5.100  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A             500      200000  thrpt    5  45849.079 ±  3332.665  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A            4000      200000  thrpt    5  41066.999 ±  4879.069  ops/s
BenchSimLJMC3D.integratorStep                    N/A             500      200000  thrpt    5  37785.229 ±  3590.728  ops/s
BenchSimLJMC3D.integratorStep                    N/A            4000      200000  thrpt    5  37072.682 ±   770.917  ops/s
BenchSimLJMD3D.integratorStep                    N/A             N/A         N/A  thrpt    5  17908.626 ± 27059.591  ops/s
BenchSimSWChain.integratorStep                   N/A             500      100000  thrpt    5    212.403 ±   128.132  ops/s
BenchSimSWChain.integratorStep                   N/A            4000      100000  thrpt    5     22.510 ±     4.642  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A            1000         N/A  thrpt    5    128.359 ±     9.479  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A           20000         N/A  thrpt    5      5.235 ±     0.053  ops/s
```

branch
```
Benchmark                                  (nCellsZ)  (numMolecules)  (numSteps)   Mode  Cnt      Score      Error  Units
BenchSimCatalysis.integratorStep                  20             N/A         N/A  thrpt    3   3956.147 ± 1217.730  ops/s
BenchSimDCVGCMD.integratorStep                   N/A             N/A         N/A  thrpt    5   1754.109 ±   78.499  ops/s
BenchSimHSMD3D.integratorStep                    N/A             500    20000000  thrpt    5   4841.402 ±  238.208  ops/s
BenchSimHSMD3D.integratorStep                    N/A            4000    20000000  thrpt    5    453.133 ±   26.465  ops/s
BenchSimHSMD3D.integratorStep                    N/A           32000    20000000  thrpt    5     40.509 ±   11.586  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A             500      200000  thrpt    5  58394.886 ± 3897.919  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A            4000      200000  thrpt    5  44358.921 ± 9279.973  ops/s
BenchSimLJMC3D.integratorStep                    N/A             500      200000  thrpt    5  48945.788 ± 2731.879  ops/s
BenchSimLJMC3D.integratorStep                    N/A            4000      200000  thrpt    5  40610.186 ± 8132.475  ops/s
BenchSimLJMD3D.integratorStep                    N/A             N/A         N/A  thrpt    5     31.724 ±   15.834  ops/s
BenchSimSWChain.integratorStep                   N/A             500      100000  thrpt    5    247.941 ±   20.867  ops/s
BenchSimSWChain.integratorStep                   N/A            4000      100000  thrpt    5     22.656 ±    3.043  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A            1000         N/A  thrpt    5    152.890 ±    4.579  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A           20000         N/A  thrpt    5      6.088 ±    0.046  ops/s
```

difference
```
Benchmark                                  (nCellsZ)  (numMolecules)  (numSteps)   Mode  Cnt      Score      Error  Units
BenchSimCatalysis.integratorStep                  20             N/A         N/A  thrpt    3   [-3533.346-]   {+3956.147+} ±   [-490.563-] {+1217.730+}  ops/s
BenchSimDCVGCMD.integratorStep                   N/A             N/A         N/A  thrpt    5   [-1571.728-]   {+1754.109+} ±    [-83.091-]   {+78.499+}  ops/s
BenchSimHSMD3D.integratorStep                    N/A             500    20000000  thrpt    5   [-4392.060-]   {+4841.402+} ±   [-107.071-]  {+238.208+}  ops/s
BenchSimHSMD3D.integratorStep                    N/A            4000    20000000  thrpt    5    [-428.471-]    {+453.133+} ±    [-39.539-]   {+26.465+}  ops/s
BenchSimHSMD3D.integratorStep                    N/A           32000    20000000  thrpt    5     [-37.111-]     {+40.509+} ±     [-5.100-]   {+11.586+}  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A             500      200000  thrpt    5  [-45849.079-]  {+58394.886+} ±  [-3332.665-] {+3897.919+}  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A            4000      200000  thrpt    5  [-41066.999-]  {+44358.921+} ±  [-4879.069-] {+9279.973+}  ops/s
BenchSimLJMC3D.integratorStep                    N/A             500      200000  thrpt    5  [-37785.229-]  {+48945.788+} ±  [-3590.728-] {+2731.879+}  ops/s
BenchSimLJMC3D.integratorStep                    N/A            4000      200000  thrpt    5  [-37072.682-]  {+40610.186+} ±   [-770.917-] {+8132.475+}  ops/s
BenchSimLJMD3D.integratorStep                    N/A             N/A         N/A  thrpt    5  [-17908.626-]     {+31.724+} ± [-27059.591-]   {+15.834+}  ops/s
BenchSimSWChain.integratorStep                   N/A             500      100000  thrpt    5    [-212.403-]    {+247.941+} ±   [-128.132-]   {+20.867+}  ops/s
BenchSimSWChain.integratorStep                   N/A            4000      100000  thrpt    5     [-22.510-]     {+22.656+} ±     [-4.642-]    {+3.043+}  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A            1000         N/A  thrpt    5    [-128.359-]    {+152.890+} ±     [-9.479-]    {+4.579+}  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A           20000         N/A  thrpt    5      [-5.235-]      {+6.088+} ±     [-0.053-]    {+0.046+}  ops/s
```

**Comment by akofke** [comment]
branch with `etomica.nbr.parallel=true` set, for reference

```
Benchmark                                  (nCellsZ)  (numMolecules)  (numSteps)   Mode  Cnt      Score       Error  Units
BenchSimCatalysis.integratorStep                  20             N/A         N/A  thrpt    3   5652.412 ± 11991.397  ops/s
BenchSimDCVGCMD.integratorStep                   N/A             N/A         N/A  thrpt    5   2048.988 ±   102.616  ops/s
BenchSimHSMD3D.integratorStep                    N/A             500    20000000  thrpt    5   6308.351 ±   148.081  ops/s
BenchSimHSMD3D.integratorStep                    N/A            4000    20000000  thrpt    5    616.163 ±    37.596  ops/s
BenchSimHSMD3D.integratorStep                    N/A           32000    20000000  thrpt    5     53.219 ±    15.510  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A             500      200000  thrpt    5  55718.233 ±  1309.264  ops/s
BenchSimLJGCMC3D.integratorStep                  N/A            4000      200000  thrpt    5  45590.461 ± 11465.907  ops/s
BenchSimLJMC3D.integratorStep                    N/A             500      200000  thrpt    5  45311.877 ±  2850.957  ops/s
BenchSimLJMC3D.integratorStep                    N/A            4000      200000  thrpt    5  41113.008 ±  5813.843  ops/s
BenchSimLJMD3D.integratorStep                    N/A             N/A         N/A  thrpt    5     52.288 ±     4.295  ops/s
BenchSimSWChain.integratorStep                   N/A             500      100000  thrpt    5    333.065 ±    16.488  ops/s
BenchSimSWChain.integratorStep                   N/A            4000      100000  thrpt    5     29.124 ±     4.034  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A            1000         N/A  thrpt    5    154.053 ±     1.248  ops/s
BenchSimSurfacetensionLJMC.integratorStep        N/A           20000         N/A  thrpt    5      6.052 ±     0.216  ops/s
```

**Comment by ajschult** [comment]
I ran the tinderbox benchmarks on on my workstation (Haswell, 2014), a slow workstation (Core2 Quad, vintage 2008) and a CCR node (Xeon Nehalem, vintage 2009).  Overall, Haswell is more than 50% faster than the Core2 Quad, and the Core2 Quad is more than 50% faster than the Xeon.

The Haswell and Core2 Quad both show 0-10% speedup across all tests (MD and MC).  The Xeon has 4-15% speedup for MD and -2 to +19% speedup for MC.  For MC 500 atoms goes faster with the branch and 4000 atoms has the same or slower speed.

The Xeon is ~40% faster than douglas.

**Comment by ajschult** [comment]
> Move ApiAACell to private inner class of BenchApiAACell 
> Since we don't need it anymore.

Do we need a benchmark for an unused class?

**Comment by ajschult** [comment]
> Add neighbor list updates benchmark 

Is this supposed to benchmark checking for neighbor updates or benchmark actually updating the neighbors?

**Comment by akofke** [comment]
> Do we need a benchmark for an unused class? 

Probably not but it does serve to compare the new vs old methods using other profilers such as GC, and as a baseline for future improvements to the cell iteration.

`BenchNbrUpdates` is supposed to measure the time to update neighbor lists in isolation. It's kind of contrived, since the update is idempotent it measures single-shot time and just uses a ton of atoms, and moves them to a random location between iterations. So not a "typical" scenario but it should measure the raw speed (taking into account data structures and synchronization unlike the cell iterator bench) in order to compare e.g. parallel vs not and using downlists vs not.

**Review by ajschult — APPROVED** [review]
Approved!

**Pull request merged**